### PR TITLE
KAFKA-16908: Refactor QuorumConfig with AbstractConfig

### PR DESCRIPTION
**Pull request by johnnychhsu**
- extend AbstractConfig for QuorumConfig
- replace previous constructor with the new one
Jira: https://issues.apache.org/jira/browse/KAFKA-16908

## Test
run `./gradlew clean raft:test --tests RaftEventSimulationTest` and passed
run `./gradlew clean build -x test` and succeed

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java [hunk 1]
```diff
@@ -108,63 +109,36 @@ public class QuorumConfig {
...
-        this.electionTimeoutMs = electionTimeoutMs;
-        this.electionBackoffMaxMs = electionBackoffMaxMs;
-        this.fetchTimeoutMs = fetchTimeoutMs;
-        this.appendLingerMs = appendLingerMs;
+        super(configDef(), props);
     }
 
     public int requestTimeoutMs() {
-        return requestTimeoutMs;
+        return getInt(QUORUM_REQUEST_TIMEOUT_MS_CONFIG);
```
Comment (OmniaGM) [inline raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java]: If we are defining these getters here can we add `QuorumConfig` to `KafkaConfig` in similar way to `_remoteLogManagerConfig` and delete the quorum getters from there (and maybe some of the validators as well).  As this will help keeping KafkaConfig tidier cc: @chia7712

**Review by OmniaGM — COMMENTED** [review]
Nice refactor, I just left one suggestion to keep things in line with refactoring KafkaConfig.

File: raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java (hunk 1 above)
Comment (chia7712) [reply raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java]: I'm re-thinking the design about inheritance and composition.  `_remoteLogManagerConfig` has a issue that it can't reflect the updated configs (#16394) and I left a comment (https://github.com/apache/kafka/pull/16394#discussion_r1646273869) about keeping a `KafkaConfig` reference instead of copying all props of `KafkaConfig`. 

It seems all configs in `QuorumConfig` can't be updated dynamically, so it is fine for now. However, I'd like to see more consistent design for those individual config class. Hence, maybe we can take a look at #16394 first, and then this PR can follow the same pattern :)

**Comment by chia7712** [comment]
@johnnychhsu Could you please take a look at #16394? It adopts another approach to complete the refactor.

File: raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java (hunk 1 above)
Comment (johnnychhsu) [reply raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java]: thanks @OmniaGM @chia7712 
let me check https://github.com/apache/kafka/pull/16394 first

Comment (johnnychhsu) [reply raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java]: just check https://github.com/apache/kafka/pull/16394 and it looks like currently the approach is instead of inheriting, we composite an AbstractConfig as a property of the actual config. 
I am wondering if in the future we can put all config related implementation in a module, instead of scattering around in different modules. In this case, take this QuorumConfig for example, we can have a constructor that accept KafkaConfig to init the QuorumConfig.

Comment (OmniaGM) [reply raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java]: I believe this is the discussion @chia7712 and I are having in https://github.com/apache/kafka/pull/16394 specially that we are trying to pull KafkaConfig out of core in other PRs.

File: core/src/main/scala/kafka/raft/RaftManager.scala [hunk 2]
```diff
@@ -157,9 +155,16 @@ class KafkaRaftManager[T](
...
+  props.put(QuorumConfig.QUORUM_REQUEST_TIMEOUT_MS_CONFIG, config.getInt(QuorumConfig.QUORUM_REQUEST_TIMEOUT_MS_CONFIG))
+  props.put(QuorumConfig.QUORUM_RETRY_BACKOFF_MS_CONFIG, config.getInt(QuorumConfig.QUORUM_RETRY_BACKOFF_MS_CONFIG))
+  props.put(QuorumConfig.QUORUM_ELECTION_TIMEOUT_MS_CONFIG, config.getInt(QuorumConfig.QUORUM_ELECTION_TIMEOUT_MS_CONFIG))
+  props.put(QuorumConfig.QUORUM_ELECTION_BACKOFF_MAX_MS_CONFIG, config.getInt(QuorumConfig.QUORUM_ELECTION_BACKOFF_MAX_MS_CONFIG))
+  props.put(QuorumConfig.QUORUM_FETCH_TIMEOUT_MS_CONFIG, config.getInt(QuorumConfig.QUORUM_FETCH_TIMEOUT_MS_CONFIG))
+  props.put(QuorumConfig.QUORUM_LINGER_MS_CONFIG, config.getInt(QuorumConfig.QUORUM_LINGER_MS_CONFIG))
 
   val apiVersions = new ApiVersions()
-  private val raftConfig = new QuorumConfig(config)
+  private val raftConfig = new QuorumConfig(props)
```
Comment (chia7712) [inline core/src/main/scala/kafka/raft/RaftManager.scala:167]: please fix the build error

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 3]
```diff
@@ -721,6 +721,8 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
   val quorumLingerMs = getInt(QuorumConfig.QUORUM_LINGER_MS_CONFIG)
   val quorumRequestTimeoutMs = getInt(QuorumConfig.QUORUM_REQUEST_TIMEOUT_MS_CONFIG)
   val quorumRetryBackoffMs = getInt(QuorumConfig.QUORUM_RETRY_BACKOFF_MS_CONFIG)
+  private val _quorumConfig = new QuorumConfig(this)
+  def quorumConfig = _quorumConfig
```
Comment (chia7712) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:680]: Could you please use this to return quorum-related configs and then remove all related getters from `KafkaConfig`? for example: `quorumElectionBackoffMs`, `quorumFetchTimeoutMs`, `quorumElectionTimeoutMs`, `quorumLingerMs` and `quorumRetryBackoffMs` can be removed. `quorumRequestTimeoutMs` can be removed and its' callers should use `QuorumConfig#requestTimeoutMs` instead

**Review by chia7712 — COMMENTED** [review]
@johnnychhsu thanks for updated PR

File: raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java [hunk 4]
```diff
@@ -99,7 +99,14 @@ public class QuorumConfig {
...
     public static final int DEFAULT_QUORUM_RETRY_BACKOFF_MS = 20;
 
-    public static final ConfigDef CONFIG_DEF =  new ConfigDef()
+    private final AbstractConfig config;
+
+    public QuorumConfig(AbstractConfig config) {
+        this.config = config;
+    }
+
+    public static ConfigDef configDef() {
```
Comment (chia7712) [inline raft/src/main/java/org/apache/kafka/raft/QuorumConfig.java:108]: Why we need this change?

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java [hunk 5]
```diff
@@ -79,6 +79,7 @@
 import java.util.Optional;
 import java.util.OptionalInt;
 import java.util.OptionalLong;
+import java.util.Properties;
```
Comment (chia7712) [inline raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java:82]: please remove unnecessary change

File: raft/src/test/java/org/apache/kafka/raft/RaftEventSimulationTest.java [hunk 6]
```diff
@@ -762,14 +763,14 @@ void start(int nodeId) {
...
-                LINGER_MS
-            );
+            Properties props = new Properties();
+            props.put(QuorumConfig.QUORUM_REQUEST_TIMEOUT_MS_CONFIG, REQUEST_TIMEOUT_MS);
+            props.put(QuorumConfig.QUORUM_RETRY_BACKOFF_MS_CONFIG, RETRY_BACKOFF_MS);
+            props.put(QuorumConfig.QUORUM_ELECTION_TIMEOUT_MS_CONFIG, ELECTION_TIMEOUT_MS);
+            props.put(QuorumConfig.QUORUM_ELECTION_BACKOFF_MAX_MS_CONFIG, ELECTION_JITTER_MS);
+            props.put(QuorumConfig.QUORUM_FETCH_TIMEOUT_MS_CONFIG, FETCH_TIMEOUT_MS);
+            props.put(QuorumConfig.QUORUM_LINGER_MS_CONFIG, LINGER_MS);
+            QuorumConfig quorumConfig = new QuorumConfig(props);
```
Comment (chia7712) [inline raft/src/test/java/org/apache/kafka/raft/RaftEventSimulationTest.java:773]: please fix compile error

**Review by chia7712 — COMMENTED** [review]
@johnnychhsu thanks for updated patch. It would be nice to check the build  :)

**Comment by chia7712** [comment]
close as this was fixed by #17231
